compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DeadCodeAnalyzer.java

```diff
@@ -0,0 +1,299 @@
+import org.wso2.ballerinalang.compiler.semantics.model.symbols.BSymbol;
+import org.wso2.ballerinalang.compiler.semantics.model.symbols.UsedState;
+import org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody;
+import org.wso2.ballerinalang.compiler.tree.BLangClassDefinition;
+import org.wso2.ballerinalang.compiler.tree.BLangFunction;
+import org.wso2.ballerinalang.compiler.tree.BLangNode;
+import org.wso2.ballerinalang.compiler.tree.BLangPackage;
+import org.wso2.ballerinalang.compiler.tree.BLangSimpleVariable;
+import org.wso2.ballerinalang.compiler.tree.SimpleBLangNodeAnalyzer;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangArrowFunction;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangInvocation;
+import org.wso2.ballerinalang.compiler.tree.expressions.BLangSimpleVarRef;
+import org.wso2.ballerinalang.compiler.tree.statements.BLangStatement;
+import org.wso2.ballerinalang.compiler.util.CompilerContext;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+
+import static org.ballerinalang.model.tree.NodeKind.CLASS_DEFN;
+import static org.ballerinalang.model.tree.NodeKind.FUNCTION;
+
+/**
+ * Analyze function invocation chains and detect unused functions
+ */
+public class DeadCodeAnalyzer extends SimpleBLangNodeAnalyzer<DeadCodeAnalyzer.AnalyzerData> {
+
+    private static final CompilerContext.Key<DeadCodeAnalyzer> DEAD_CODE_ANALYZER_KEY = new CompilerContext.Key<>();
+    final DeadCodeAnalyzer.AnalyzerData data = new DeadCodeAnalyzer.AnalyzerData();
+
+    private DeadCodeAnalyzer(CompilerContext context) {
+
+        context.put(DEAD_CODE_ANALYZER_KEY, this);
+    }
+
+    public static DeadCodeAnalyzer getInstance(CompilerContext context) {
+
+        DeadCodeAnalyzer deadCodeAnalyzer = context.get(DEAD_CODE_ANALYZER_KEY);
+        if (deadCodeAnalyzer == null) {
+            deadCodeAnalyzer = new DeadCodeAnalyzer(context);
+        }
+        return deadCodeAnalyzer;
+    }
+
+    public BLangPackage analyze(BLangPackage pkgNode) {
+
+        visitNode(pkgNode, data);
+        return pkgNode;
+    }
+
+    private void analyzeTopLevelNodes(BLangPackage pkgNode, DeadCodeAnalyzer.AnalyzerData data) {
+
+        List<TopLevelNode> topLevelNodes = pkgNode.topLevelNodes;
+
+        // Identifying all available functions and registering lambda function pointers
+        for (TopLevelNode topLevelNode : topLevelNodes) {
+            if (topLevelNode.getKind() == FUNCTION) {
+                BLangFunction topLevelFunction = (BLangFunction) topLevelNode;
+                if (isLambda(topLevelFunction)) {
+                    addLambdaFunctionToMap(topLevelFunction);
+                } else {
+                    addFunctionToMap(topLevelFunction);
+                }
+            }
+            if (topLevelNode.getKind() == CLASS_DEFN) {
+                for (BLangFunction classLevelFunction : ((BLangClassDefinition) topLevelNode).functions) {
```

Comment:
We can wrap `(BLangClassDefinition) topLevelNode` this to a new variable

Reply:
Modified in commit #a0abba10a989aee9ad4c0ab14369cbeff55c8907